File: 175_invert-binary-tree/invert-binary-tree.java
/*
@Copyright:LintCode
@Author:   taoleetju
@Problem:  http://www.lintcode.com/problem/invert-binary-tree
@Language: Java
@Datetime: 15-12-04 14:44
*/

/**
 * Definition of TreeNode:
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left, right;
 *     public TreeNode(int val) {
 *         this.val = val;
 *         this.left = this.right = null;
 *     }
 * }
 */
public class Solution {
    /**
     * @param root: a TreeNode, the root of the binary tree
     * @return: nothing
     */
    public void invertBinaryTree(TreeNode root) {
        TreeNode tmp = root.left;
        root.left= root.right;
        root.right = tmp;
        if (root.left != null)
        this.invertBinaryTree(root.left);
        if (root.right != null)
        this.invertBinaryTree(root.right);
    }
}

File: 175_invert-binary-tree/invert-binary-tree.cpp
/*
@Copyright:LintCode
@Author:   taoleetju
@Problem:  http://www.lintcode.com/problem/invert-binary-tree
@Language: C++
@Datetime: 16-05-23 16:59
*/

/**
 * Definition of TreeNode:
 * class TreeNode {
 * public:
 *     int val;
 *     TreeNode *left, *right;
 *     TreeNode(int val) {
 *         this->val = val;
 *         this->left = this->right = NULL;
 *     }
 * }
 */
 #include <deque>
 using namespace std;
class Solution {
public:
    /**
     * @param root: a TreeNode, the root of the binary tree
     * @return: nothing
     */
    void invertBinaryTree(TreeNode *root) {
        // write your code here
        /*
        //递归版本
        if(root ==NULL)
            return;
        TreeNode *temp;
        temp = root->left;
        root->left = root->right;
        root->right = temp;
        invertBinaryTree(root->left);
        invertBinaryTree(root->right);
        */
        //非递归版本
        if(root == NULL)
            return;
        deque<TreeNode *> d;
        d.push_back(root);
        TreeNode *front, *temp;
        while(!d.empty()){
            front = d.front();
            if(front->left!=NULL)
                d.push_back(front->left);
            if(front->right!=NULL)
                d.push_back(front->right);
            temp = front->left;
            front->left = front->right;
            front->right =temp;
            d.pop_front();
        }
    }
};